### PR TITLE
L1 Deposit Tx

### PR DESCRIPTION
**Pull request by NiloCK**
PR is partial work toward an e2e deposit workflow. It:

- modifies `e2e.Stack.Run` to return some stack data for the test runner
- crafts and submits a deposit tx to the L1 deposit contract, via the op-bindings `portal`
- asserts that the tx landed on L1

The deposit tx is not being processed by Monomer. Likely because monomer's epoch does not advance, in turn likely related to some rollback issues.

To do (future PRs):
- [ ] assert that deposits are picked up from L1 by the node
- [ ] assert minting / balance changes on L2
- [ ] pack a deposit tx with transaction data, processed and executed by monomer

A concern:
- `stack_test.go` is becoming a little unwieldy. May be worth a focused refactor at some point to extract some sort of `action,result` table test so that causes and observable effects can be coupled together more effectively.

## Summary by CodeRabbit

- **New Features**
  - Enhanced the stack's functionality to return detailed configuration information alongside error handling.
  - Introduced a new `StackConfig` structure for better management of configuration parameters.
  - Integrated Layer 1 interactions for deposit transactions in end-to-end tests, including transaction signing and verification.

- **Bug Fixes**
  - Improved error management in the stack setup process.

### CODE REVIEW

File: e2e/stack_test.go
Comment (natebeauregard) [inline e2e/stack_test.go:217]: Why is this sleep call necessary now?

File: e2e/stack_test.go
Comment (natebeauregard) [inline e2e/stack_test.go:202]: Is it worth asserting any information on the deposit tx receipt? It might be nice to verify that the `TransactionDeposited` event was emitted

File: e2e/stack_test.go
Comment (natebeauregard) [inline e2e/stack_test.go:131]: Why do you fallback to the original price here? Should the max gas price just be used instead?

File: e2e/stack_test.go
Comment (natebeauregard) [inline e2e/stack_test.go:152]: Readbility might be better if you extract `big.NewInt(1e18/20)` out to a contant

File: e2e/stack_test.go
Comment (natebeauregard) [inline e2e/stack_test.go:145]: Could you use the `gasPrice` variable you initialized above here?

File: e2e/stack_test.go
Comment (natebeauregard) [inline e2e/stack_test.go:146]: Why is `1e6` used here for gas limit? I see that the gas limit is also passed in as `stackConfig.Genesis.SystemConfig.GasLimit/10` down below

**Review by natebeauregard — COMMENTED** [review]
Changes look real good! Left a couple questions for my own understanding but I'll approve once they're answered

File: e2e/stack.go
Comment (joshklop) [inline e2e/stack.go:205]: I haven't looked through the PR in detail yet, but my impression is that this can be quite dangerous. The caller will be able to modify the data provided in this `StackConfig`. Probably better for the caller to pass in this kind of information to the `Stack`.

Comment (joshklop) [reply e2e/stack.go:205]: On second thought, my points don't make much sense. I'll let Nate take the reviews here.

File: e2e/stack_test.go
Comment (NiloCK) [reply e2e/stack_test.go:202]: This suggestion bore fruit. As it stood, the transaction was reverting. Event verification added here: 5253de9. Adjustment to MaxGas that fixed the revert in the prior commit.

File: e2e/stack_test.go
Comment (NiloCK) [reply e2e/stack_test.go]: b169ea0, ec36f24

These are mostly dummy values. I extracted `oneEth` and used it directly.

File: e2e/stack_test.go
Comment (NiloCK) [reply e2e/stack_test.go]: Covered in https://github.com/polymerdao/monomer/commit/c65a3dfb7478eb3edec8e6b74b2d09db9dc90638. Redundant gas price calculation removed.

File: e2e/stack_test.go
Comment (NiloCK) [reply e2e/stack_test.go]: Mostly trial and error. The `1e6` here is the gasLimit for the L1 tx. It has since changed to `1e7`, which fixes a revert.

The other gasLimit is for L2 processing of the deposit transaction.

Comment (natebeauregard) [reply e2e/stack_test.go]: Ah ok thanks. Could you pull it out into a constant for readability?

File: e2e/stack_test.go
Comment (natebeauregard) [inline e2e/stack_test.go:232]: Can you remove this comment before merge?

**Review by natebeauregard — APPROVED** [review]
LGTM! Left a couple small suggestions for a constant to be extracted and for a comment to be removed but no need to re-review before squash/merge

File: e2e/stack_test.go
Comment (NiloCK) [reply e2e/stack_test.go]: Refactored in 71d7221, but there's the interplay between L1 and L2 gasLimit values that discourages me from pulling either as a constant.

(L1 gas gets burned according to the l2Gaslimit as part of the cross-chain gas accounting.)

File: e2e/stack_test.go
Comment (NiloCK) [reply e2e/stack_test.go]: Always sneaks back in. I just love sleeping for 15 seconds.

File: e2e/stack.go
Comment (NiloCK) [reply e2e/stack.go:205]: Third thought - I think these points do make sense. The returned data should be "information only", and not knobs of the running stack that can be twiddled by the caller.

Tracked in [notion](https://www.notion.so/nethermind/e2e-return-copied-data-from-stack-Run-db0e763ab0d8489598574f718e3d8edd?pvs=4).

**Pull request merged**
